### PR TITLE
Add authentication module

### PR DESCRIPTION
**Pull request by kisom**
This PR adds support for an authenticated signing endpoint and a default HMAC authentication provider. This extends the work originally done by @grittygrease.

### CODE REVIEW

**Comment by grittygrease** [comment]
A couple more fixes for the comments above and this should be good to go.

File: signer/standard_signer.go [hunk 1]
```diff
@@ -199,7 +198,7 @@ func (s *StandardSigner) sign(template *x509.Certificate, profile *config.Signin
 		}
 	}
 	cert = pem.EncodeToMemory(&pem.Block{Type: "CERTIFICATE", Bytes: derBytes})
-	log.Infof("Signed certificate with serial number %s", serialNumber)
+	log.Infof("signed certificate with serial number %s", serialNumber)
```
Comment (kisom) [inline signer/standard_signer.go]: I think we should look at this in another branch. It requires some extra work that isn't orthogonal to this PR.

File: api/client/client.go [hunk 2]
```diff
@@ -43,14 +49,68 @@ func (srv *Server) getURL(endpoint string) string {
...
+	}
+
+	buf := bytes.NewBuffer(jsonData)
+	resp, err := http.Post(url, "application/json", buf)
+	if err != nil {
+		return nil, err
+	}
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
```
Comment (grittygrease) [inline api/client/client.go]: Wrap ioutil error with CFSSL wrapper.

File: signer/standard_signer.go [hunk 3]
```diff
@@ -223,7 +225,8 @@ func (s *StandardSigner) Sign(hostName string, in []byte, subject *Subject, prof
 
 	template, err := ParseCertificateRequest(s, block.Bytes, subject)
 	if err != nil {
-		return
+		return nil, cferr.New(cferr.CertificateError,
+			cferr.DecodeFailed, errors.New("not a certificate or csr"))
```
Comment (grittygrease) [inline signer/standard_signer.go]: This error does not need to be wrapped, ParseCertificateRequest returns a cfssl error.

File: api/client/client.go [hunk 4]
```diff
@@ -43,14 +49,68 @@ func (srv *Server) getURL(endpoint string) string {
...
+
+	jsonData, err := json.Marshal(aReq)
+	if err != nil {
+		return nil, err
+	}
+
+	buf := bytes.NewBuffer(jsonData)
+	resp, err := http.Post(url, "application/json", buf)
+	if err != nil {
+		return nil, err
```
Comment (kisom) [inline api/client/client.go]: The API client now has its own category in errors.

**Comment by grittygrease** [comment]
Looks good.

File: api/api_test.go [hunk 5]
```diff
@@ -20,7 +20,7 @@ const (
 )
 
 func newTestSignHandler(t *testing.T) (h http.Handler) {
-	h, err := NewSignHandler(testCaFile, testCaKeyFile, "")
+	h, err := NewSignHandler(testCaFile, testCaKeyFile, "", nil)
```
Comment (lziest) [inline api/api_test.go]: I understand the test needs a (mock) remote signer. Ignore the above comment. Let's address this in another PR.

**Comment by lziest** [comment]
LGTM

File: api/api_test.go (hunk 5 above)
Comment (kisom) [inline api/api_test.go]: Currently working on a testing approach for this now.

**Pull request merged**